librosa/core/spectrum.py

```diff
@@ -519,6 +520,450 @@ def ifgram(y, sr=22050, n_fft=2048, hop_length=None, win_length=None,
+
+        See `stft` for details.
+
+        .. see also:: `filters.get_window`
+
+    ref_power : float >= 0 or callable
+        Minimum power threshold for estimating time-frequency reassignments.
+        Any bin with `np.abs(S[f, t])**2 < ref_power` will be returned as
+        `np.nan` in both frequency and time.
+
+    clip : boolean
+        - If `True`, estimated frequencies outside the range `[0, 0.5 * sr]`
+          or times outside the range `[0, len(y) / sr]` will be returned as
+          `np.nan`.
+        - If `False`, estimated frequencies and times beyond the domain and
+          range of the spectrogram may be returned.
+
+    dtype : numeric type
+        Complex numeric type for `S`. Default is 64-bit complex.
+
+    Returns
+    -------
+    freqs : np.ndarray [shape=(1 + n_fft/2, t), dtype=real]
+        Instantaneous frequencies:
+        `freqs[f, t]` is the frequency for bin `f`, frame `t`
+
+    times : np.ndarray [shape=(1 + n_fft/2, t), dtype=real]
+        Reassigned times:
+        `times[f, t]` is the time for bin `f`, frame `t`
+
+    mags : np.ndarray [shape=(1 + n_fft/2, t), dtype=real]
+        Magnitudes from short-time Fourier transform:
+        `mags[f, t]` is the magnitude for bin `f`, frame `t`
+
+    Warns
+    --------
+    RuntimeWarning
+        Frequency or time estimates with zero support will produce a
+        divide-by-zero warning and will be returned as `np.nan`.
+
+    See Also
+    --------
+    stft : Short-time Fourier Transform
+    reassign_frequencies : Estimate frequency reassignments
+    reassign_times : Estimate time reassignments
+
+    Examples
+    --------
+    >>> y, sr = librosa.load(librosa.util.example_audio_file())
+    >>> y_zoom = y[27 * sr : 29 * sr]
+
+    >>> freqs, times, mags = librosa.reassigned_spectrogram(
+    ...     y=y_zoom, sr=sr, hop_length=16, n_fft=64, ref_power=1e-4
+    ... )
+    >>> db = librosa.amplitude_to_db(mags, ref=np.max)
+
+    >>> import matplotlib.pyplot as plt
+    >>> plt.figure()
+    >>> plt.subplot(2, 1, 1)
+    >>> librosa.display.specshow(
+    ...     db, x_axis="s", y_axis="linear", sr=sr, hop_length=16
+    ... )
+    >>> plt.title("Spectrogram")
+    >>> plt.subplot(2, 1, 2)
+    >>> plt.scatter(times, freqs, c=db, s=0.1, cmap="magma")
+    >>> plt.title("Reassigned spectrogram")
+    >>> plt.xlim([0, 2])
+    >>> plt.xticks([0, 0.5, 1, 1.5])
+    >>> plt.ylabel("Hz")
+    >>> plt.subplots_adjust(
+    ...     left=0.1, bottom=0.05, right=0.95, top=0.95, hspace=0.5
+    ... )
+    >>> plt.show()
+
+    """
+
+    freqs, S = reassign_frequencies(
+        y,
+        sr=sr,
+        n_fft=n_fft,
+        hop_length=hop_length,
+        win_length=win_length,
+        window=window,
+        dtype=dtype,
+    )
+
+    times, S = reassign_times(
+        y,
+        sr=sr,
+        n_fft=n_fft,
+        hop_length=hop_length,
+        win_length=win_length,
+        window=window,
+        dtype=dtype,
+    )
+
```

Comment:
There's a lot of redundant calculation here.  Can we refactor this to improve efficiency?

Reply:
I'm not sure about how to improve the redundant calculation of the basic STFT. This algorithm requires three STFTs: the basic one (`S_h`), plus one calculated with the window derivative (`S_dh`, for frequency reassignment) and one calculated with a time-weighted window (`S_th`, for time reassignment).

Two other ways that I thought about organizing `reassigned_spectrogram` so that the basic STFT is only calculated once:
- Put the calculation of both auxiliary windows and all three STFTs into `reassigned_spectrogram` (and probably remove the two duplicate support functions). This disadvantages are that it would be a big function and tests would be harder to write. For users who only want frequency reassignment or time reassignment, this function would calculate one unneeded STFT.
- Allow the two support functions to accept the original STFT as input. In `reassigned_spectrogram`, the STFT would only be calculated in the first function, and then passed to the other. My thoughts on this are in the first PR comment.

What do you think?

Reply:
>     * Allow the two support functions to accept the original STFT as input.

I like this option, and don't see any obvious downsides to it.  I agree that 3 different stft's are necessary, but we're doing 4 now, and that's my main objection.

Reply:
Done in https://github.com/librosa/librosa/pull/926/commits/30695ee22631138fbc10207738c5421f65bd1be6